R/tm_g_ae_oview.R

```diff
@@ -285,34 +281,32 @@ srv_g_ae_oview <- function(id,
 
     output_q <- reactive({
       ANL <- data[[dataname]]() # nolint
-      validate(need(iv$is_valid(), "Misspecification error: please observe red flags in the encodings."))
-      validate(need(
-        is.factor(ANL[[input$arm_var]]),
-        "Selected arm variable needs to be a factor."
-      ))
-      validate(need(input$flag_var_anl, "Please select at least one flag."))
-      iv_comp <- shinyvalidate::InputValidator$new()
-      iv_comp$add_rule("arm_trt", shinyvalidate::sv_not_equal(
+
+      teal::validate_has_data(ANL, min_nrow = 10, msg = sprintf("%s has not enough data", dataname))
+
+      # set up and enable input validator(s)
+      iv <- shinyvalidate::InputValidator$new()
```

Comment:
Does the validator need to be created inside a reactive - couldn't it be created in the module and then just call gather_fails() in the reactive?

Reply:
In this particular module, yes. This is because some inputs are created in response to others. See the original version, there were actually two* validators, one created at the top of the module (line 235) and the other one in the reactive (line 294).

In some modules this is not a requirement, but in some it is a must. I decided to keep it in the reactive in all modules for consistency.

* I honestly didn't check for the third one at 309

Reply:
So this works outside of the reactive:

```
    # set up and enable input validator(s)
    iv <- shinyvalidate::InputValidator$new()
    iv$add_rule("arm_var", shinyvalidate::sv_required(
      message = "Arm Variable is required"))
    iv$add_rule("arm_var", ~ if (!is.factor(data[[dataname]]()[[req(.)]]))
      "Arm Var must be a factor variable")
    iv$add_rule("arm_var", ~ if (length(unique(data[[dataname]]()[[req(.)]])) < 2)
      "Selected Arm Var has not enough treatments to compare")
    iv$add_rule("flag_var_anl", shinyvalidate::sv_required(
      message = "At least one Flag is required"))
    iv$add_rule("arm_trt", ~ if (!is.null(input$arm_ref) && (.) == input$arm_ref)
      message_fmt = "Control and Treatment must be different")
    iv$add_rule("arm_ref", ~ if (!is.null(input$arm_trt) && (.) == input$arm_trt)
      message_fmt = "Control and Treatment must be different")
    iv$enable()
```

But it is a little more painful...

Reply:
So, three things.

1. It seems the source of the problem is the `shinyvalidate` helper functions and all we have to do is avoid them and we will be fine. Good catch!

2. This 
```
iv$add_rule("arm_trt", ~ if (!is.null(input$arm_ref) && (.) == input$arm_ref)
```
can be replaced with this
```
iv$add_rule("arm_trt", ~ if (isTRUE(. == input$arm_ref))
```
`isTRUE` and `isFALSE` saves us the trouble of testing `!is.null(.)`.

3 Unfortunately, this fails if placed outside ot the reactive:
```
iv$add_rule("arm_trt", shinyvalidate::sv_in_set(
  set = unique(data[[dataname]]()[[input$arm_var]]),
  message_fmt = "Treatment not found in Arm Variable"))
```
and so does this:
```
iv$add_rule("arm_trt", ~ if (isFALSE(. %in% unique(data[[dataname]]()[[input$arm_var]])))
  "Treatment not found in Arm Variable")
```

Reply:
Also, when the validator is outside of the reactive, red messages appear on the widgets while the UI is being rendered, before the firs cycle is complete. Would we be fine with that? This could perhaps be prevented by adding a condition to the validator, but I'm not sure. And adding "all inputs are not NULL" is a) again, bulky and unseemly, b) may derail optional inputs.

Reply:
Interesting - I think it's not worth moving stuff around in teal.osprey to make it work outside of the reactive - but thought it worth exploring.

I would have expected the second part of part 3. to work but 🤷 

I think the extra reactive cycle (I assume due to our use of insertUI - yet again a problem @gogonzo ) is probably what causes the red errors - if that could be fied in teal at somepoint then maybe we'd explore moving the validator outside of the reactive.

For tmc where there is a much more rigid structure, it's probably worth deciding exactly where to put the validator and to do it in the same place in all the modules

Reply:
I think the main reason why things break down outside the reactive is the `updateSelectInput`s, but I have yet to confirm it.

The extra cycle may just be because the second input is updated after the first one is fully generated (a default value is properly selected). This also happens when one generates inputs server side (`selectInput` in `renderUI` and another `selectInput` in another `renderUI` dependent on that). I don't think we would be able to eliminate the extra cycles completely but putting the validator in the reactive does prevent the red  flashes (in this case).

Speaking of extra cycles, in some modules (not sure about `oview`) if you put a `browser` at the beginning of the reactive it will be triggered three times before the app fully starts. Probably nothing to worry about yet, but something to keep in mind.

Reply:
And yes, `~ if (!is.null(input$arm_var) && !((.) %in% unique(data[[dataname]]()[[input$arm_var]])))` does work. The `!is.null` prevents a subseting error that would derail `isFALSE`. Hats off, sir.

Reply:
Having the validation independent of reactive calls is very desirable and I dont mind the red errors upon initialization but I understand the concerns here.